calyx-opt/src/passes/static_promotion.rs

```diff
@@ -523,31 +642,86 @@ impl Visitor for StaticPromotion {
     fn finish_seq(
         &mut self,
         s: &mut ir::Seq,
-        _comp: &mut ir::Component,
-        _sigs: &LibrarySignatures,
+        comp: &mut ir::Component,
+        sigs: &LibrarySignatures,
         _comps: &[ir::Component],
     ) -> VisResult {
-        if let Some(sseq) = s.make_static() {
-            return Ok(Action::change(ir::Control::Static(
-                ir::StaticControl::Seq(sseq),
-            )));
+        let mut builder = ir::Builder::new(comp, sigs);
+        let mut new_stmts: Vec<ir::Control> = Vec::new();
+        let mut static_vec: Vec<ir::Control> = Vec::new();
+        let mut promote_control = true;
+        for stmt in std::mem::take(&mut s.stmts) {
+            if stmt.is_static()
+                || stmt.has_attribute(ir::NumAttr::PromoteStatic)
+            {
+                static_vec.push(stmt);
+            } else {
+                promote_control = false;
+                match static_vec.len().cmp(&1) {
+                    std::cmp::Ordering::Equal => {
+                        new_stmts.extend(static_vec);
+                    }
+                    std::cmp::Ordering::Greater => {
+                        let sseq = self.construct_static_seq(
+                            &mut builder,
+                            &mut static_vec,
+                        );
+                        new_stmts.push(sseq);
+                    }
+                    _ => {}
+                }
+                new_stmts.push(stmt);
+                static_vec = Vec::new();
```

Comment:
Instead of setting `static_vec = Vec::new()` at the end of each iteration, you could instead pass `static_vec.drain(..).collect` as the argument to `construct_static_seq`.

Reply:
Inactionable. This doesn't pass cargo check on my computer?